src/chat.cpp

```diff
@@ -1131,6 +1226,215 @@ CCMD( say_team )
 	}
 }
 
+//*****************************************************************************
+//
+// [AK] Checks if the player is eligible to receive private messages from us.
+//
+bool chat_IsPlayerValidReceiver( ULONG ulPlayer )
+{
+	// [AK] If the player doesn't exist, they can't be a receiver.
+	if ( PLAYER_IsValidPlayer( ulPlayer ) == false )
+		return false;
+
+	// [AK] The receiver can't be ourselves or a bot.
+	if (( ulPlayer == static_cast<ULONG>( consoleplayer )) || ( players[ulPlayer].bIsBot ))
+		return false;
+
+	return true;
+}
+
+//*****************************************************************************
+//
+// [AK] Finds a valid player whom we can send private messages to.
+//
+void chat_FindValidReceiver( void )
+{
+	bool nobodyFound = true;
+
+	if ( SERVER_CountPlayers( false ) >= 2 )
+	{
+		// [AK] We can always send private messages to the server.
+		if ( g_ulChatPlayer == MAXPLAYERS )
+			return;
+
+		nobodyFound = false;
+
+		// [AK] If we're trying to send a message to an invalid player, find another one.
+		if ( chat_IsPlayerValidReceiver( g_ulChatPlayer ) == false )
+		{
+			ULONG oldPlayer = g_ulChatPlayer;
+
+			do
+			{
+				// [AK] Keep looping until we find another player who we can send a message to.
+				if ( ++g_ulChatPlayer >= MAXPLAYERS )
+					g_ulChatPlayer = 0;
+
+				// [AK] If we're back to the old value for some reason, then there's no other
+				// players to send a message to, so set the receiver to the server instead.
+				if ( g_ulChatPlayer == oldPlayer )
+				{
+					nobodyFound = true;
+					break;
+				}
+			}
+			while ( chat_IsPlayerValidReceiver( g_ulChatPlayer ) == false );
+		}
+	}
+
+	if ( nobodyFound )
+	{
+		Printf( "There's no other clients to send private messages to. "
+			"The server will be selected instead.\n" );
+		g_ulChatPlayer = MAXPLAYERS;
+	}
+}
+
+//*****************************************************************************
+//
+// [AK] Allows players (or the server) to send private messages to other players.
+//
+void chat_PrivateMessage( FCommandLine &argv, const ULONG ulReceiver, const bool bServerSelected )
+{

```

Comment:
Why do we need a separate argument `bServerSelected` to encode the server as target instead of just using `ulReceiver == MAXPLAYERS`?
BTW: Sorry that this comment only shows up now. I posted it last week but unfortunately using "Start a review" instead of "Add single comment". I just now realized that you couldn't see it.

Reply:
Thank you for asking. I added `bServerSelected` to ignore all the extra checks required to validate another player being the receiver, if the client explicitly chooses to the server. If the client wants indicates that they want to send a private message to the server using `sayto server` or `sayto_idx -1`, then it's not necessary to check if the receiver is invalid, a bot, ourselves, etc.

Additionally, if `bServerSelected` is true, then the function `chat_FindValidReceiver` won't be entered at all. If we did enter that function and there was only one client connected, then the message "There's no other clients to send private messages to. The server will be selected instead." would always appear in the console, even if the client intentionally chose to send a private message to the server.

Reply:
You could also bypass all those checks by using `ulReceiver == MAXPLAYERS` instead of `bServerSelected`. For instance, you could add `const bool bServerSelected = (ulReceiver == MAXPLAYERS);` in the first line of the implementation of `chat_PrivateMessage` and keep the rest of `chat_PrivateMessage` untouched (except for removing the argument `bServerSelected`). Perhaps there is a reason why you don't want to encode the information `bServerSelected` in the argument `ulReceiver`, but right now I don't see it.

Reply:
Fair enough, I can change that quickly and resubmit it if you're still around.

Reply:
Yes, I'm still around.

Reply:
Sorry, but I gotta go now. I should be able to add this tomorrow though.

Reply:
No worries, sorry for the wait. I just around to adding that last patch, but if you can still add it in tomorrow that's awesome!